{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7b4948d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.\n",
      "Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv('your/path/RBD_valid_nucleotides_million.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "69da5b60",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession ID</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>RBD nucleotide</th>\n",
       "      <th>Variant</th>\n",
       "      <th>RBD protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EPI_ISL_3197453</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...</td>\n",
       "      <td>alpha</td>\n",
       "      <td>RVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Accession ID</td>\n",
       "      <td>Lineage</td>\n",
       "      <td>RBD nucleotide</td>\n",
       "      <td>Variant</td>\n",
       "      <td>RBD protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EPI_ISL_16973626</td>\n",
       "      <td>BA.5.2.1</td>\n",
       "      <td>AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...</td>\n",
       "      <td>omicron</td>\n",
       "      <td>RVQPTESIVRFPNITNLCPFDEVFNATRFASVYAWNRKRISNCVAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Accession ID</td>\n",
       "      <td>Lineage</td>\n",
       "      <td>RBD nucleotide</td>\n",
       "      <td>Variant</td>\n",
       "      <td>RBD protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EPI_ISL_2090530</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...</td>\n",
       "      <td>alpha</td>\n",
       "      <td>RVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVAD...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Accession ID   Lineage  \\\n",
       "0   EPI_ISL_3197453   B.1.1.7   \n",
       "1      Accession ID   Lineage   \n",
       "2  EPI_ISL_16973626  BA.5.2.1   \n",
       "3      Accession ID   Lineage   \n",
       "4   EPI_ISL_2090530   B.1.1.7   \n",
       "\n",
       "                                      RBD nucleotide  Variant  \\\n",
       "0  AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...    alpha   \n",
       "1                                     RBD nucleotide  Variant   \n",
       "2  AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...  omicron   \n",
       "3                                     RBD nucleotide  Variant   \n",
       "4  AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...    alpha   \n",
       "\n",
       "                                         RBD protein  \n",
       "0  RVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVAD...  \n",
       "1                                        RBD protein  \n",
       "2  RVQPTESIVRFPNITNLCPFDEVFNATRFASVYAWNRKRISNCVAD...  \n",
       "3                                        RBD protein  \n",
       "4  RVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVAD...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "92a3e7d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[df['Accession ID'].str.startswith('EPI_')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "844aa8a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession ID</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>RBD nucleotide</th>\n",
       "      <th>Variant</th>\n",
       "      <th>RBD protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EPI_ISL_3197453</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...</td>\n",
       "      <td>alpha</td>\n",
       "      <td>RVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EPI_ISL_16973626</td>\n",
       "      <td>BA.5.2.1</td>\n",
       "      <td>AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...</td>\n",
       "      <td>omicron</td>\n",
       "      <td>RVQPTESIVRFPNITNLCPFDEVFNATRFASVYAWNRKRISNCVAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EPI_ISL_2090530</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...</td>\n",
       "      <td>alpha</td>\n",
       "      <td>RVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>EPI_ISL_13391748</td>\n",
       "      <td>BA.2</td>\n",
       "      <td>AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...</td>\n",
       "      <td>omicron</td>\n",
       "      <td>RVQPTESIVRFPNITNLCPFDEVFNATRFASVYAWNRKRISNCVAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>EPI_ISL_14789048</td>\n",
       "      <td>BA.5.2</td>\n",
       "      <td>AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...</td>\n",
       "      <td>omicron</td>\n",
       "      <td>RVQPTESIVRFPNITNLCPFDEVFNATRFASVYAWNRKRISNCVAD...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Accession ID   Lineage  \\\n",
       "0   EPI_ISL_3197453   B.1.1.7   \n",
       "2  EPI_ISL_16973626  BA.5.2.1   \n",
       "4   EPI_ISL_2090530   B.1.1.7   \n",
       "6  EPI_ISL_13391748      BA.2   \n",
       "8  EPI_ISL_14789048    BA.5.2   \n",
       "\n",
       "                                      RBD nucleotide  Variant  \\\n",
       "0  AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...    alpha   \n",
       "2  AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...  omicron   \n",
       "4  AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...    alpha   \n",
       "6  AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...  omicron   \n",
       "8  AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...  omicron   \n",
       "\n",
       "                                         RBD protein  \n",
       "0  RVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVAD...  \n",
       "2  RVQPTESIVRFPNITNLCPFDEVFNATRFASVYAWNRKRISNCVAD...  \n",
       "4  RVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVAD...  \n",
       "6  RVQPTESIVRFPNITNLCPFDEVFNATRFASVYAWNRKRISNCVAD...  \n",
       "8  RVQPTESIVRFPNITNLCPFDEVFNATRFASVYAWNRKRISNCVAD...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d1dbb04e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.loc[:, 'Variant'] = df['Variant'].replace('Not a VOC', 'nonVOC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d8a6bc8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('your/path/RBD_valid_nucleotides.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fc4507a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No 'N' found in the 'RBD nucleotide' column.\n"
     ]
    }
   ],
   "source": [
    "if df['RBD nucleotide'].str.contains('N').any():\n",
    "    print(\"There is at least one 'N' in the 'RBD nucleotide' column.\")\n",
    "else:\n",
    "    print(\"No 'N' found in the 'RBD nucleotide' column.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "36cbc432",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Lineage\n",
      "B.1.1.7      118481\n",
      "BA.2          67107\n",
      "AY.4          50307\n",
      "BA.1.1        48689\n",
      "BA.2.12.1     25196\n",
      "              ...  \n",
      "B.1.1.361         1\n",
      "EP.1.1            1\n",
      "B.1.1.265         1\n",
      "B.1.1.112         1\n",
      "B.1.250           1\n",
      "Name: count, Length: 3155, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the CSV file\n",
    "df = pd.read_csv(\"your/path/RBD_valid_nucleotides.csv\")\n",
    "\n",
    "# Count the number of rows for each unique value in the \"Lineage\" column\n",
    "lineage_counts = df['Lineage'].value_counts()\n",
    "\n",
    "# Print the counts\n",
    "print(lineage_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "12585b0d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession ID</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>RBD nucleotide</th>\n",
       "      <th>Variant</th>\n",
       "      <th>RBD protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EPI_ISL_402124</td>\n",
       "      <td>A.1</td>\n",
       "      <td>AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...</td>\n",
       "      <td>WuhanHu1</td>\n",
       "      <td>RVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVAD...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Accession ID Lineage                                     RBD nucleotide  \\\n",
       "0  EPI_ISL_402124     A.1  AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...   \n",
       "\n",
       "    Variant                                        RBD protein  \n",
       "0  WuhanHu1  RVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVAD...  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df[\"Accession ID\"]==\"EPI_ISL_402124\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "63bd1c1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# creating sample subset for testing:\n",
    "df = pd.read_csv('your/path/RBD_valid_nucleotides.csv')\n",
    "subset_df = df.head(1000)\n",
    "subset_df.to_csv('subset_df.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "f92f219e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Lineage\n",
      "B.1.1.7      110\n",
      "BA.2          66\n",
      "BA.1.1        50\n",
      "AY.4          43\n",
      "B.1.617.2     24\n",
      "            ... \n",
      "BA.5.2.20      1\n",
      "BA.5.9         1\n",
      "BA.2.3.17      1\n",
      "BF.21          1\n",
      "AY.93          1\n",
      "Name: count, Length: 263, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "lineage_counts = subset_df['Lineage'].value_counts()\n",
    "\n",
    "# Print the counts\n",
    "print(lineage_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "87c5b164",
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Virus name</th>\n",
       "      <th>Last vaccinated</th>\n",
       "      <th>Passage details/history</th>\n",
       "      <th>Type</th>\n",
       "      <th>Accession ID</th>\n",
       "      <th>Collection date</th>\n",
       "      <th>Location</th>\n",
       "      <th>Additional location information</th>\n",
       "      <th>Sequence length</th>\n",
       "      <th>Host</th>\n",
       "      <th>...</th>\n",
       "      <th>Pango version</th>\n",
       "      <th>Variant</th>\n",
       "      <th>AA Substitutions</th>\n",
       "      <th>Submission date</th>\n",
       "      <th>Is reference?</th>\n",
       "      <th>Is complete?</th>\n",
       "      <th>Is high coverage?</th>\n",
       "      <th>Is low coverage?</th>\n",
       "      <th>N-Content</th>\n",
       "      <th>GC-Content</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hCoV-19/India/KA-RFNB-6196/2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Original</td>\n",
       "      <td>betacoronavirus</td>\n",
       "      <td>EPI_ISL_10746595</td>\n",
       "      <td>2022-02</td>\n",
       "      <td>Asia / India / Karnataka</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29884</td>\n",
       "      <td>Human</td>\n",
       "      <td>...</td>\n",
       "      <td>consensus call</td>\n",
       "      <td>Former VOC Omicron GRA (B.1.1.529+BA.*) first ...</td>\n",
       "      <td>(NSP5_P132H,NSP3_G489S,NSP4_T327I,Spike_S373P,...</td>\n",
       "      <td>2022-03-07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.378998795342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hCoV-19/India/KA-RFNB-6116/2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Original</td>\n",
       "      <td>betacoronavirus</td>\n",
       "      <td>EPI_ISL_10746594</td>\n",
       "      <td>2022-02</td>\n",
       "      <td>Asia / India / Karnataka</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29884</td>\n",
       "      <td>Human</td>\n",
       "      <td>...</td>\n",
       "      <td>consensus call</td>\n",
       "      <td>Former VOC Omicron GRA (B.1.1.529+BA.*) first ...</td>\n",
       "      <td>(NSP5_P132H,Spike_S371F,NSP3_G489S,NSP13_R392C...</td>\n",
       "      <td>2022-03-07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.379166108955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hCoV-19/India/KA-RFNB-6166/2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Original</td>\n",
       "      <td>betacoronavirus</td>\n",
       "      <td>EPI_ISL_10746593</td>\n",
       "      <td>2022-02</td>\n",
       "      <td>Asia / India / Karnataka</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29902</td>\n",
       "      <td>Human</td>\n",
       "      <td>...</td>\n",
       "      <td>consensus call</td>\n",
       "      <td>Former VOC Omicron GRA (B.1.1.529+BA.*) first ...</td>\n",
       "      <td>(NSP5_P132H,Spike_S371F,NSP3_G489S,NSP13_R392C...</td>\n",
       "      <td>2022-03-07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.3792388469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hCoV-19/India/KA-RFNB-6141/2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Original</td>\n",
       "      <td>betacoronavirus</td>\n",
       "      <td>EPI_ISL_10746592</td>\n",
       "      <td>2022-02</td>\n",
       "      <td>Asia / India / Karnataka</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29884</td>\n",
       "      <td>Human</td>\n",
       "      <td>...</td>\n",
       "      <td>consensus call</td>\n",
       "      <td>Former VOC Omicron GRA (B.1.1.529+BA.*) first ...</td>\n",
       "      <td>(NSP5_P132H,NSP3_G489S,NSP13_R392C,Spike_K417N...</td>\n",
       "      <td>2022-03-07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.379032258065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hCoV-19/India/KA-RFNB-6175/2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Original</td>\n",
       "      <td>betacoronavirus</td>\n",
       "      <td>EPI_ISL_10746591</td>\n",
       "      <td>2022-02</td>\n",
       "      <td>Asia / India / Karnataka</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29893</td>\n",
       "      <td>Human</td>\n",
       "      <td>...</td>\n",
       "      <td>consensus call</td>\n",
       "      <td>Former VOC Omicron GRA (B.1.1.529+BA.*) first ...</td>\n",
       "      <td>(NSP5_P132H,NSP3_G489S,NSP4_T327I,Spike_S373P,...</td>\n",
       "      <td>2022-03-07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.379051951962</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Virus name Last vaccinated Passage details/history  \\\n",
       "0  hCoV-19/India/KA-RFNB-6196/2022             NaN                Original   \n",
       "1  hCoV-19/India/KA-RFNB-6116/2022             NaN                Original   \n",
       "2  hCoV-19/India/KA-RFNB-6166/2022             NaN                Original   \n",
       "3  hCoV-19/India/KA-RFNB-6141/2022             NaN                Original   \n",
       "4  hCoV-19/India/KA-RFNB-6175/2022             NaN                Original   \n",
       "\n",
       "              Type      Accession ID Collection date  \\\n",
       "0  betacoronavirus  EPI_ISL_10746595         2022-02   \n",
       "1  betacoronavirus  EPI_ISL_10746594         2022-02   \n",
       "2  betacoronavirus  EPI_ISL_10746593         2022-02   \n",
       "3  betacoronavirus  EPI_ISL_10746592         2022-02   \n",
       "4  betacoronavirus  EPI_ISL_10746591         2022-02   \n",
       "\n",
       "                   Location Additional location information Sequence length  \\\n",
       "0  Asia / India / Karnataka                             NaN           29884   \n",
       "1  Asia / India / Karnataka                             NaN           29884   \n",
       "2  Asia / India / Karnataka                             NaN           29902   \n",
       "3  Asia / India / Karnataka                             NaN           29884   \n",
       "4  Asia / India / Karnataka                             NaN           29893   \n",
       "\n",
       "    Host  ...   Pango version  \\\n",
       "0  Human  ...  consensus call   \n",
       "1  Human  ...  consensus call   \n",
       "2  Human  ...  consensus call   \n",
       "3  Human  ...  consensus call   \n",
       "4  Human  ...  consensus call   \n",
       "\n",
       "                                             Variant  \\\n",
       "0  Former VOC Omicron GRA (B.1.1.529+BA.*) first ...   \n",
       "1  Former VOC Omicron GRA (B.1.1.529+BA.*) first ...   \n",
       "2  Former VOC Omicron GRA (B.1.1.529+BA.*) first ...   \n",
       "3  Former VOC Omicron GRA (B.1.1.529+BA.*) first ...   \n",
       "4  Former VOC Omicron GRA (B.1.1.529+BA.*) first ...   \n",
       "\n",
       "                                    AA Substitutions Submission date  \\\n",
       "0  (NSP5_P132H,NSP3_G489S,NSP4_T327I,Spike_S373P,...      2022-03-07   \n",
       "1  (NSP5_P132H,Spike_S371F,NSP3_G489S,NSP13_R392C...      2022-03-07   \n",
       "2  (NSP5_P132H,Spike_S371F,NSP3_G489S,NSP13_R392C...      2022-03-07   \n",
       "3  (NSP5_P132H,NSP3_G489S,NSP13_R392C,Spike_K417N...      2022-03-07   \n",
       "4  (NSP5_P132H,NSP3_G489S,NSP4_T327I,Spike_S373P,...      2022-03-07   \n",
       "\n",
       "  Is reference? Is complete? Is high coverage? Is low coverage? N-Content  \\\n",
       "0           NaN         True               NaN              NaN       NaN   \n",
       "1           NaN         True               NaN              NaN       NaN   \n",
       "2           NaN         True               NaN              NaN       NaN   \n",
       "3           NaN         True               NaN              NaN       NaN   \n",
       "4           NaN         True               NaN              NaN       NaN   \n",
       "\n",
       "       GC-Content  \n",
       "0  0.378998795342  \n",
       "1  0.379166108955  \n",
       "2    0.3792388469  \n",
       "3  0.379032258065  \n",
       "4  0.379051951962  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7ee0e1ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Python script that will take in csv file that contains RBD nuc sequences and protein sequences to compare \n",
    "# Accession ID values with metadata.tsv to access Collection date value for each row and divide each Lineage \n",
    "# group into 10 Generations unique to the lineage based on Collection date range\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# Load the CSV files\n",
    "# metadata_df = pd.read_csv(\"metadata.tsv\", sep='\\t', dtype=str, low_memory=False, encoding='latin-1')\n",
    "nuc_prot_df = pd.read_csv(\"your/path/RBD_valid_nucleotides.csv\")\n",
    "\n",
    "# Merge the two dataframes on Accession ID\n",
    "merged_df = nuc_prot_df.merge(metadata_df[['Accession ID', 'Collection date']], on='Accession ID', how='left')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "1cd505e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession ID</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>RBD nucleotide</th>\n",
       "      <th>Variant</th>\n",
       "      <th>RBD protein</th>\n",
       "      <th>Collection date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EPI_ISL_402124</td>\n",
       "      <td>A.1</td>\n",
       "      <td>AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...</td>\n",
       "      <td>WuhanHu1</td>\n",
       "      <td>RVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVAD...</td>\n",
       "      <td>2019-12-30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Accession ID Lineage                                     RBD nucleotide  \\\n",
       "0  EPI_ISL_402124     A.1  AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...   \n",
       "\n",
       "    Variant                                        RBD protein Collection date  \n",
       "0  WuhanHu1  RVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVAD...      2019-12-30  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df.loc[merged_df[\"Accession ID\"]==\"EPI_ISL_402124\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "74904b11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession ID</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>RBD nucleotide</th>\n",
       "      <th>Variant</th>\n",
       "      <th>RBD protein</th>\n",
       "      <th>Collection date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EPI_ISL_402124</td>\n",
       "      <td>A.1</td>\n",
       "      <td>AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...</td>\n",
       "      <td>WuhanHu1</td>\n",
       "      <td>RVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVAD...</td>\n",
       "      <td>2019-12-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EPI_ISL_3197453</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...</td>\n",
       "      <td>alpha</td>\n",
       "      <td>RVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVAD...</td>\n",
       "      <td>2021-04-28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EPI_ISL_16973626</td>\n",
       "      <td>BA.5.2.1</td>\n",
       "      <td>AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...</td>\n",
       "      <td>omicron</td>\n",
       "      <td>RVQPTESIVRFPNITNLCPFDEVFNATRFASVYAWNRKRISNCVAD...</td>\n",
       "      <td>2023-02-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EPI_ISL_2090530</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...</td>\n",
       "      <td>alpha</td>\n",
       "      <td>RVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVAD...</td>\n",
       "      <td>2021-04-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EPI_ISL_13391748</td>\n",
       "      <td>BA.2</td>\n",
       "      <td>AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...</td>\n",
       "      <td>omicron</td>\n",
       "      <td>RVQPTESIVRFPNITNLCPFDEVFNATRFASVYAWNRKRISNCVAD...</td>\n",
       "      <td>2022-02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Accession ID   Lineage  \\\n",
       "0    EPI_ISL_402124       A.1   \n",
       "1   EPI_ISL_3197453   B.1.1.7   \n",
       "2  EPI_ISL_16973626  BA.5.2.1   \n",
       "3   EPI_ISL_2090530   B.1.1.7   \n",
       "4  EPI_ISL_13391748      BA.2   \n",
       "\n",
       "                                      RBD nucleotide   Variant  \\\n",
       "0  AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...  WuhanHu1   \n",
       "1  AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...     alpha   \n",
       "2  AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...   omicron   \n",
       "3  AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...     alpha   \n",
       "4  AGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAA...   omicron   \n",
       "\n",
       "                                         RBD protein Collection date  \n",
       "0  RVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVAD...      2019-12-30  \n",
       "1  RVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVAD...      2021-04-28  \n",
       "2  RVQPTESIVRFPNITNLCPFDEVFNATRFASVYAWNRKRISNCVAD...      2023-02-01  \n",
       "3  RVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVAD...      2021-04-25  \n",
       "4  RVQPTESIVRFPNITNLCPFDEVFNATRFASVYAWNRKRISNCVAD...         2022-02  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "678306f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output CSV file saved as RBD_valid_Gen.csv\n"
     ]
    }
   ],
   "source": [
    "# Convert Collection date to %Y-%m-%d format\n",
    "merged_df['Collection date'] = pd.to_datetime(merged_df['Collection date'], format='%Y-%m-%d', errors='coerce')\n",
    "merged_df['Collection date'] = merged_df['Collection date'].dt.strftime('%Y-%m-%d')\n",
    "\n",
    "# Filter rows with Collection date after November 2019\n",
    "merged_df['Collection date'] = pd.to_datetime(merged_df['Collection date'])\n",
    "merged_df = merged_df[merged_df['Collection date'] >= '2019-11-01']\n",
    "\n",
    "# Group data by Lineage\n",
    "grouped_data = merged_df.groupby('Lineage')\n",
    "\n",
    "# Define a function to calculate 'Generation' for each Lineage group\n",
    "def assign_generation(group):\n",
    "    date_range = group['Collection date'].max() - group['Collection date'].min()\n",
    "    bin_size = date_range / 10\n",
    "    group['Generation'] = pd.cut(group['Collection date'], bins=10, labels=['Gen1', 'Gen2', 'Gen3', 'Gen4', 'Gen5', 'Gen6', 'Gen7', 'Gen8', 'Gen9', 'Gen10'])\n",
    "    return group\n",
    "\n",
    "\n",
    "# Filter Lineage groups with more than 10 sequences\n",
    "filtered_groups = [group for name, group in grouped_data if len(group) > 6]\n",
    "\n",
    "# Apply the assign_generation function to each filtered Lineage group\n",
    "processed_data = pd.concat([assign_generation(group) for group in filtered_groups])\n",
    "\n",
    "# Save the updated data to a new CSV file\n",
    "output_csv = \"your/path/RBD_valid_Gen.csv\"\n",
    "processed_data.to_csv(output_csv, index=False)\n",
    "\n",
    "print(f\"Output CSV file saved as {output_csv}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "153c7c60",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession ID</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>RBD nucleotide</th>\n",
       "      <th>RBD protein</th>\n",
       "      <th>Variant</th>\n",
       "      <th>Collection date</th>\n",
       "      <th>Generation</th>\n",
       "      <th>Dateframe</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>EPI_ISL_411956</td>\n",
       "      <td>A</td>\n",
       "      <td>AATGGAACCATTACAGATGCTGTAGACTGTGCACTTGACCCTCTCT...</td>\n",
       "      <td>NGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIV...</td>\n",
       "      <td>WuhanHu1</td>\n",
       "      <td>2020-02-11</td>\n",
       "      <td>Gen1</td>\n",
       "      <td>Dateframe1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>EPI_ISL_520744</td>\n",
       "      <td>A</td>\n",
       "      <td>AATGGAACCATTACAGATGCTGTAGACTGTGCACTTGACCCTCTCT...</td>\n",
       "      <td>NGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIV...</td>\n",
       "      <td>WuhanHu1</td>\n",
       "      <td>2020-04-21</td>\n",
       "      <td>Gen2</td>\n",
       "      <td>Dateframe2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>724</th>\n",
       "      <td>EPI_ISL_17682083</td>\n",
       "      <td>A</td>\n",
       "      <td>AATGGAACCATTACAGATGCTGTAGACTGTGCACTTGACCCTCTCT...</td>\n",
       "      <td>NGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIV...</td>\n",
       "      <td>WuhanHu1</td>\n",
       "      <td>2020-06-22</td>\n",
       "      <td>Gen2</td>\n",
       "      <td>Dateframe2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6942</th>\n",
       "      <td>EPI_ISL_776445</td>\n",
       "      <td>A</td>\n",
       "      <td>AATGGAACCATTACAGATGCTGTAGACTGTGCACTTGACCCTCTCT...</td>\n",
       "      <td>NGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIV...</td>\n",
       "      <td>WuhanHu1</td>\n",
       "      <td>2020-03-28</td>\n",
       "      <td>Gen1</td>\n",
       "      <td>Dateframe1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7759</th>\n",
       "      <td>EPI_ISL_901221</td>\n",
       "      <td>A</td>\n",
       "      <td>AATGGAACCATTACAGATGCTGTAGACTGTGCACTTGACCCTCTCT...</td>\n",
       "      <td>NGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIV...</td>\n",
       "      <td>WuhanHu1</td>\n",
       "      <td>2020-03-09</td>\n",
       "      <td>Gen1</td>\n",
       "      <td>Dateframe1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Accession ID Lineage  \\\n",
       "124     EPI_ISL_411956       A   \n",
       "211     EPI_ISL_520744       A   \n",
       "724   EPI_ISL_17682083       A   \n",
       "6942    EPI_ISL_776445       A   \n",
       "7759    EPI_ISL_901221       A   \n",
       "\n",
       "                                         RBD nucleotide  \\\n",
       "124   AATGGAACCATTACAGATGCTGTAGACTGTGCACTTGACCCTCTCT...   \n",
       "211   AATGGAACCATTACAGATGCTGTAGACTGTGCACTTGACCCTCTCT...   \n",
       "724   AATGGAACCATTACAGATGCTGTAGACTGTGCACTTGACCCTCTCT...   \n",
       "6942  AATGGAACCATTACAGATGCTGTAGACTGTGCACTTGACCCTCTCT...   \n",
       "7759  AATGGAACCATTACAGATGCTGTAGACTGTGCACTTGACCCTCTCT...   \n",
       "\n",
       "                                            RBD protein   Variant  \\\n",
       "124   NGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIV...  WuhanHu1   \n",
       "211   NGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIV...  WuhanHu1   \n",
       "724   NGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIV...  WuhanHu1   \n",
       "6942  NGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIV...  WuhanHu1   \n",
       "7759  NGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIV...  WuhanHu1   \n",
       "\n",
       "     Collection date Generation   Dateframe  \n",
       "124       2020-02-11       Gen1  Dateframe1  \n",
       "211       2020-04-21       Gen2  Dateframe2  \n",
       "724       2020-06-22       Gen2  Dateframe2  \n",
       "6942      2020-03-28       Gen1  Dateframe1  \n",
       "7759      2020-03-09       Gen1  Dateframe1  "
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "processed_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "9a1fae05",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the RBD_valid_Gen.csv file\n",
    "input_file = \"your/path/RBD_valid_Gen.csv\"\n",
    "df = pd.read_csv(input_file)\n",
    "\n",
    "# Subset the DataFrame to contain only rows with \"BA.1\" in the \"Variant\" column\n",
    "subset_df = df[df[\"Lineage\"] == \"BA.1\"]\n",
    "subset_df.to_csv('your/path/subset.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bc33ff0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38e6cbe4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c042f5a8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "261236e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the input CSV file\n",
    "input_file = \"your/path/RBD_valid_Gen.csv\"\n",
    "df = pd.read_csv(input_file)\n",
    "\n",
    "# Initialize an empty list to store DataFrames\n",
    "output_dfs = []\n",
    "\n",
    "# Group the data by Lineage\n",
    "grouped = df.groupby(\"Lineage\")\n",
    "\n",
    "# Iterate through each Lineage group\n",
    "for lineage, group in grouped:\n",
    "    generations = group[\"Generation\"].unique()\n",
    "    generations = sorted([int(gen.replace(\"Gen\", \"\")) for gen in generations])  # Extract and convert to int\n",
    "\n",
    "    # Filter for consecutive Generation values\n",
    "    for i in range(len(generations) - 1):\n",
    "        gen1, gen2 = generations[i], generations[i + 1]\n",
    "        if abs(gen2 - gen1) == 1:\n",
    "            # Get the rows with Generation values gen1 and gen2\n",
    "            rows_gen1 = group[group[\"Generation\"] == f\"Gen{gen1}\"]\n",
    "            rows_gen2 = group[group[\"Generation\"] == f\"Gen{gen2}\"]\n",
    "\n",
    "            # Check if there are rows for the specified generations\n",
    "            if not rows_gen1.empty and not rows_gen2.empty:\n",
    "                # Create a new DataFrame for the output\n",
    "                min_len = min(len(rows_gen1), len(rows_gen2))\n",
    "                new_df = pd.DataFrame({\n",
    "                    \"Lineage\": [lineage] * min_len,\n",
    "                    \"GenS1\": [f\"Gen{gen1}\"] * min_len,\n",
    "                    \"GenS2\": [f\"Gen{gen2}\"] * min_len,\n",
    "                    \n",
    "                    #\"Seq1\": rows_gen1[\"RBD protein\"].values[:min_len],\n",
    "                    #\"Seq2\": rows_gen2[\"RBD protein\"].values[:min_len],\n",
    "\n",
    "                    \"Seq1\": rows_gen1[\"RBD nucleotide\"].values[:min_len],\n",
    "                    \"Seq2\": rows_gen2[\"RBD nucleotide\"].values[:min_len],\n",
    "                    \"Variant\": rows_gen1[\"Variant\"].values[:min_len],\n",
    "                })\n",
    "\n",
    "                # Append the new DataFrame to the list\n",
    "                output_dfs.append(new_df)\n",
    "\n",
    "# Concatenate all DataFrames in the list into one output DataFrame\n",
    "output_df = pd.concat(output_dfs, ignore_index=True)\n",
    "\n",
    "# Save the output DataFrame to a CSV file\n",
    "#output_file = \"RBD_valid_finetuning_proteins.csv\"\n",
    "output_file = \"your/path/RBD_valid_finetuning_nucleotides.csv\"\n",
    "output_df.to_csv(output_file, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9ef3b9d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
